File: ios/Classes/MapViewPlugin.m
#import "MapViewPlugin.h"
#import "MapViewController.h"
#import "MapAnnotation.h"
#import "MapPolyline.h"
#import "MapPolygon.h"
#import <GoogleMaps/GoogleMaps.h>

@implementation MapViewPlugin

+ (void)registerWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *channel = [FlutterMethodChannel
            methodChannelWithName:@"com.apptreesoftware.map_view"
                  binaryMessenger:[registrar messenger]];
    UIViewController *host = UIApplication.sharedApplication.delegate.window.rootViewController;
    MapViewPlugin *instance = [[MapViewPlugin alloc] initWithHost:host channel:channel registrar:registrar];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)initWithHost:(UIViewController *)host channel:(FlutterMethodChannel *)channel registrar:(NSObject <FlutterPluginRegistrar> *)registrar {
    if (self = [super init]) {
        self.host = host;
        self.channel = channel;
        self.registrar = registrar;
    }
    return self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([@"setApiKey" isEqualToString:call.method]) {
        [GMSServices provideAPIKey:call.arguments];
        result(@YES);
    } else if ([@"show" isEqualToString:call.method]) {
        NSDictionary *args = call.arguments;
        NSDictionary *mapOptions = args[@"mapOptions"];
        NSDictionary *cameraDict = mapOptions[@"cameraPosition"];
        self.mapTitle = mapOptions[@"title"];

        if (mapOptions[@"mapViewType"]  != (id) [NSNull null]) {
             NSString *mapViewTypeName = mapOptions[@"mapViewType"];
             int mapType = [self getMapViewType:mapViewTypeName];
             self.mapViewType = mapType;
        }

        MapViewController *vc = [[MapViewController alloc] initWithPlugin:self
                                                          navigationItems:[self buttonItemsFromActions:args[@"actions"]]
                                                           cameraPosition:[self cameraPositionFromDict:cameraDict]];
        UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:vc];
        navController.navigationBar.hidden = [mapOptions[@"hideToolbar"] boolValue];
        navController.navigationBar.translucent = NO;
        [self.host presentViewController:navController animated:true completion:nil];
        self.mapViewController = vc;
        [self.mapViewController setMapOptions:[mapOptions[@"showUserLocation"] boolValue] locationButton:[mapOptions[@"showMyLocationButton"] boolValue] compassButton:[mapOptions[@"showCompassButton"] boolValue]];
        result(@YES);
    } else if ([@"getVisibleMarkers" isEqualToString:call.method]) {
        result(self.mapViewController.visibleMarkers);
    } else if ([@"setAnnotations" isEqualToString:call.method]) {
        [self handleSetAnnotations:call.arguments];
        result(@YES);
    } else if ([@"clearAnnotations" isEqualToString:call.method]) {
        [self.mapViewController clearAnnotations];
        result(@YES);
    } else if ([@"addAnnotation" isEqualToString:call.method]) {
        [self handleAddAnnotation:call.arguments];
        result(@YES);
    } else if ([@"removeAnnotation" isEqualToString:call.method]) {
        [self handleRemoveAnnotation:call.arguments];
        result(@YES);
    } else if ([@"getVisiblePolylines" isEqualToString:call.method]) {
        result(self.mapViewController.visiblePolylines);
    } else if ([@"clearPolylines" isEqualToString:call.method]) {
        [self.mapViewController clearPolylines];
        result(@YES);
    } else if ([@"setPolylines" isEqualToString:call.method]) {
        [self handleSetPolylines:call.arguments];
        result(@YES);
    } else if ([@"addPolyline" isEqualToString:call.method]) {
        [self handleAddPolyline:call.arguments];
        result(@YES);
    } else if ([@"removePolyline" isEqualToString:call.method]) {
        [self handleRemovePolyline:call.arguments];
        result(@YES);
    } else if ([@"getVisiblePolygons" isEqualToString:call.method]) {
        result(self.mapViewController.visiblePolygons);
    } else if ([@"clearPolygons" isEqualToString:call.method]) {
        [self.mapViewController clearPolygons];
        result(@YES);
    } else if ([@"setPolygons" isEqualToString:call.method]) {
        [self handleSetPolygons:call.arguments];
        result(@YES);
    } else if ([@"addPolygon" isEqualToString:call.method]) {
        [self handleAddPolygon:call.arguments];
        result(@YES);
    } else if ([@"removePolygon" isEqualToString:call.method]) {
        [self handleRemovePolygon:call.arguments];
        result(@YES);
    } else if ([@"setCamera" isEqualToString:call.method]) {
        [self handleSetCamera:call.arguments];
        result(@YES);
    } else if ([@"zoomToFit" isEqualToString:call.method]) {
        [self.mapViewController zoomToAnnotations:[((NSNumber *) call.arguments) intValue]];
        result(@YES);
    } else if ([@"zoomToAnnotations" isEqualToString:call.method]) {
        [self handleZoomToAnnotations:call.arguments];
        result(@YES);
    } else if ([@"getCenter" isEqualToString:call.method]) {
        CLLocationCoordinate2D location = self.mapViewController.centerLocation;
        result(@{@"latitude": @(location.latitude), @"longitude": @(location.longitude)});
    } else if ([@"getZoomLevel" isEqualToString:call.method]) {
        result(@(self.mapViewController.zoomLevel));
    } else if ([@"dismiss" isEqualToString:call.method]) {
        if (self.mapViewController) {
            [self.host dismissViewControllerAnimated:true completion:nil];
        }
        [self.mapViewController shutdown];
        self.mapViewController = nil;
        result(@YES);
    } else {
        result(FlutterMethodNotImplemented);
    }
}


- (NSArray *)buttonItemsFromActions:(NSArray *)actions {
    NSMutableArray *buttons = [NSMutableArray array];
    if (actions) {
        for (NSDictionary *action in actions) {
            UIBarButtonItem *button = [[UIBarButtonItem alloc] initWithTitle:[action valueForKey:@"title"]
                                                                       style:UIBarButtonItemStylePlain
                                                                      target:self
                                                                      action:@selector(handleToolbar:)];
            button.tag = [[action valueForKey:@"identifier"] intValue];
            [buttons addObject:button];
        }
    }
    return buttons;
}

- (void)handleSetAnnotations:(NSArray *)annotations {
    NSMutableArray *array = [NSMutableArray array];
    for (NSDictionary *aDict in annotations) {
        MapAnnotation *annotation = [MapAnnotation annotationFromDictionary:aDict];
        if (annotation) {
            [array addObject:annotation];
        }
    }
    [self.mapViewController updateAnnotations:array];
}

- (void)handleAddAnnotation:(NSDictionary *)dict {
    MapAnnotation *annotation = [MapAnnotation annotationFromDictionary:dict];
    if (annotation) {
        [self.mapViewController addAnnotation:annotation];
    }
}

- (void)handleRemoveAnnotation:(NSDictionary *)dict {
    MapAnnotation *annotation = [MapAnnotation annotationFromDictionary:dict];
    if (annotation) {
        [self.mapViewController removeAnnotation:annotation];
    }
}
- (void)handleSetPolylines:(NSArray *)polylines {
    NSMutableArray *array = [NSMutableArray array];
    for (NSDictionary *aDict in polylines) {
        MapPolyline *polyline=[MapPolyline polylineFromDictionary:aDict];
        if (polyline) {
            [array addObject:polyline];
        }
    }
    [self.mapViewController updatePolylines:array];
}

- (void)handleAddPolyline:(NSDictionary *)dict {
    MapPolyline *polyline = [MapPolyline polylineFromDictionary:dict];
    if (polyline) {
        [self.mapViewController addPolyline:polyline];
    }
}

- (void)handleRemovePolyline:(NSDictionary *)dict {
    MapPolyline *polyline = [MapPolyline polylineFromDictionary:dict];
    if (polyline) {
        [self.mapViewController removePolyline:polyline];
    }
}
- (void)handleSetPolygons:(NSArray *)polygons {
    NSMutableArray *array = [NSMutableArray array];
    for (NSDictionary *aDict in polygons) {
        MapPolygon *polygon=[MapPolygon polygonFromDictionary:aDict];
        if (polygon) {
            [array addObject:polygon];
        }
    }
    [self.mapViewController updatePolygons:array];
}

- (void)handleAddPolygon:(NSDictionary *)dict {
    MapPolygon *polygon = [MapPolygon polygonFromDictionary:dict];
    if (polygon) {
        [self.mapViewController addPolygon:polygon];
    }
}

- (void)handleRemovePolygon:(NSDictionary *)dict {
    MapPolygon *polygon = [MapPolygon polygonFromDictionary:dict];
    if (polygon) {
        [self.mapViewController removePolygon:polygon];
    }
}
- (void)handleZoomToAnnotations:(NSDictionary *)zoomToDict {
    NSArray *annotations = zoomToDict[@"annotations"];
    float padding = [zoomToDict[@"padding"] floatValue];
    [self.mapViewController zoomTo:annotations padding:padding];
}

- (void)handleToolbar:(UIBarButtonItem *)item {
    [self.channel invokeMethod:@"onToolbarAction" arguments:@(item.tag)];
}

- (void)handleSetCamera:(NSDictionary *)cameraUpdate {
    CLLocationCoordinate2D coordinate = CLLocationCoordinate2DMake([cameraUpdate[@"latitude"] doubleValue], [cameraUpdate[@"longitude"] doubleValue]);
    [self.mapViewController setCamera:coordinate zoom:[cameraUpdate[@"zoom"] floatValue] bearing:[cameraUpdate[@"bearing"] doubleValue] tilt:[cameraUpdate[@"tilt"] doubleValue]];
}

- (void)onMapReady {
    [self.channel invokeMethod:@"onMapReady" arguments:nil];
}

- (void)locationDidUpdate:(CLLocation *)location {
    NSInteger time = location.timestamp.timeIntervalSince1970;
    time *= 1000;
    [self.channel invokeMethod:@"locationUpdated" arguments:@{@"latitude": @(location.coordinate.latitude),
                                                              @"longitude": @(location.coordinate.longitude),
                                                              @"time":@(time),
                                                              @"altitude":@(location.altitude),
                                                              @"speed":@(location.speed),
                                                              @"bearing":@(location.course),
                                                              @"horizontalAccuracy":@(location.horizontalAccuracy),
                                                              @"verticalAccuracy":@(location.verticalAccuracy)
                                                              }];
}

- (void)annotationTapped:(NSString *)identifier{
    [self.channel invokeMethod:@"annotationTapped" arguments:identifier];
}
- (void)annotationDragStart:(NSString *)identifier position:(CLLocationCoordinate2D)position {
    [self.channel invokeMethod:@"annotationDragStart" arguments:@{@"id": identifier, @"latitude": @(position.latitude),@"longitude": @(position.longitude)}];
}
- (void)annotationDragEnd:(NSString *)identifier position:(CLLocationCoordinate2D)position {
    [self.channel invokeMethod:@"annotationDragEnd" arguments:@{@"id": identifier, @"latitude": @(position.latitude),@"longitude": @(position.longitude)}];
}
- (void)annotationDrag:(NSString *)identifier position:(CLLocationCoordinate2D)position{
    [self.channel invokeMethod:@"annotationDrag" arguments:@{@"id": identifier, @"latitude": @(position.latitude),@"longitude": @(position.longitude)}];
}
- (void)polylineTapped:(NSString *)identifier {
    [self.channel invokeMethod:@"polylineTapped" arguments:identifier];
}
- (void)polygonTapped:(NSString *)identifier {
    [self.channel invokeMethod:@"polygonTapped" arguments:identifier];
}
- (void)infoWindowTapped:(GMSMarker *)identifier {
    [self.channel invokeMethod:@"infoWindowTapped" arguments:identifier];
}

- (void)mapTapped:(CLLocationCoordinate2D)coordinate {
    [self.channel invokeMethod:@"mapTapped" arguments:@{@"latitude": @(coordinate.latitude), @"longitude": @(coordinate.longitude)}];
}

- (void)cameraPositionChanged:(GMSCameraPosition *)position {
    [self.channel invokeMethod:@"cameraPositionChanged" arguments:@{
            @"latitude": @(position.target.latitude),
            @"longitude": @(position.target.longitude),
            @"bearing": @(position.bearing),
            @"tilt": @(position.viewingAngle),
            @"zoom": @(position.zoom)
    }];
}

- (GMSCameraPosition *)cameraPositionFromDict:(NSDictionary *)dict {
    double latitude = [dict[@"latitude"] doubleValue];
    double longitude = [dict[@"longitude"] doubleValue];
    float zoom = [dict[@"zoom"] floatValue];
    GMSCameraPosition *camera = [GMSCameraPosition cameraWithTarget:CLLocationCoordinate2DMake(latitude, longitude) zoom:zoom];
    return camera;
}

- (int)getMapViewType:(NSString *)mapViewTypeName {
    int mapType = kGMSTypeNormal;
    if ([@"none" isEqualToString:mapViewTypeName]) {
        mapType = kGMSTypeNone;
    }
    else if ([@"satellite" isEqualToString:mapViewTypeName]) {
        mapType = kGMSTypeSatellite;
    }
    else if ([@"terrain" isEqualToString:mapViewTypeName]) {
        mapType = kGMSTypeTerrain;
    }
    else if ([@"hybrid" isEqualToString:mapViewTypeName]) {
        mapType = kGMSTypeHybrid;
    }
    else if ([@"none" isEqualToString:mapViewTypeName]) {
        mapType = kGMSTypeNone;
    }
    return mapType;
}

-(NSString *) getAssetPath:(NSString *)iconPath{
    NSString* key = [self.registrar lookupKeyForAsset:iconPath];
    NSString* path= [[NSBundle mainBundle] pathForResource:key ofType:nil];
    return path;
}

@end

File: android/src/main/kotlin/com/apptreesoftware/mapview/MapViewPlugin.kt
package com.apptreesoftware.mapview

import android.app.Activity
import android.content.Intent
import android.content.res.AssetFileDescriptor
import android.location.Location
import android.os.Build
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
/*
Every time i reformatted the code, this imports were removed so i put them here
for easier access.

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
 */
class MapViewPlugin(val activity: Activity) : MethodCallHandler {
    val mapTypeMapping: HashMap<String, Int> = hashMapOf(
            "none" to GoogleMap.MAP_TYPE_NONE,
            "normal" to GoogleMap.MAP_TYPE_NORMAL,
            "satellite" to GoogleMap.MAP_TYPE_SATELLITE,
            "terrain" to GoogleMap.MAP_TYPE_TERRAIN,
            "hybrid" to GoogleMap.MAP_TYPE_HYBRID
    )

    companion object {
        lateinit var channel: MethodChannel
        var toolbarActions: List<ToolbarAction> = emptyList()
        var showUserLocation: Boolean = false
        var showMyLocationButton: Boolean = false
        var showCompassButton: Boolean = false
        var hideToolbar: Boolean = false
        var mapTitle: String = ""
        lateinit var initialCameraPosition: CameraPosition
        var mapActivity: MapActivity? = null
        val REQUEST_GOOGLE_PLAY_SERVICES = 1000
        var mapViewType: Int = GoogleMap.MAP_TYPE_NORMAL
        lateinit var registrar: Registrar

        @JvmStatic
        fun registerWith(registrar: Registrar): Unit {
            channel = MethodChannel(registrar.messenger(), "com.apptreesoftware.map_view")
            val plugin = MapViewPlugin(activity = registrar.activity())
            channel.setMethodCallHandler(plugin)
            this.registrar = registrar
        }

        fun handleToolbarAction(id: Int) {
            channel.invokeMethod("onToolbarAction", id)
        }

        fun onMapReady() {
            channel.invokeMethod("onMapReady", null)
        }

        fun getToolbarActions(actionsList: List<Map<String, Any>>?): List<ToolbarAction> {
            if (actionsList == null) return emptyList()
            val actions = ArrayList<ToolbarAction>()
            actionsList.mapTo(actions) { ToolbarAction(it) }
            return actions
        }

        fun getCameraPosition(map: Map<String, Any>): CameraPosition {
            val latitude = map["latitude"] as Double
            val longitude = map["longitude"] as Double
            val zoom = map["zoom"] as Double
            return CameraPosition(LatLng(latitude, longitude), zoom.toFloat(), 0.0f, 0.0f)
        }

        fun mapTapped(latLng: LatLng) {
            this.channel.invokeMethod("mapTapped",
                    mapOf("latitude" to latLng.latitude,
                            "longitude" to latLng.longitude))
        }

        fun annotationTapped(id: String) {
            this.channel.invokeMethod("annotationTapped", id)
        }

        fun annotationDragStart(id: String, latLng: LatLng) {
            this.channel.invokeMethod("annotationDragStart", mapOf(
                    "id" to id,
                    "latitude" to latLng.latitude,
                    "longitude" to latLng.longitude
            ))
        }

        fun annotationDragEnd(id: String, latLng: LatLng) {
            this.channel.invokeMethod("annotationDragEnd", mapOf(
                    "id" to id,
                    "latitude" to latLng.latitude,
                    "longitude" to latLng.longitude
            ))
        }

        fun annotationDrag(id: String, latLng: LatLng) {
            this.channel.invokeMethod("annotationDrag", mapOf(
                    "id" to id,
                    "latitude" to latLng.latitude,
                    "longitude" to latLng.longitude
            ))
        }

        fun polylineTapped(id: String) {
            this.channel.invokeMethod("polylineTapped", id)
        }

        fun polygonTapped(id: String) {
            this.channel.invokeMethod("polygonTapped", id)
        }

        fun cameraPositionChanged(pos: CameraPosition) {
            this.channel.invokeMethod("cameraPositionChanged", mapOf(
                    "latitude" to pos.target.latitude,
                    "longitude" to pos.target.longitude,
                    "zoom" to pos.zoom,
                    "bearing" to pos.bearing,
                    "tilt" to pos.tilt
            ))
        }

        fun locationDidUpdate(loc: Location) {
            var verticalAccuracy = 0.0f
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
                verticalAccuracy = loc.verticalAccuracyMeters
            this.channel.invokeMethod("locationUpdated", mapOf(
                    "latitude" to loc.latitude,
                    "longitude" to loc.longitude,
                    "time" to loc.time,
                    "altitude" to loc.altitude,
                    "speed" to loc.speed,
                    "bearing" to loc.bearing,
                    "horizontalAccuracy" to loc.accuracy,
                    "verticalAccuracy" to verticalAccuracy
            ))
        }

        fun infoWindowTapped(id: String) {
            this.channel.invokeMethod("infoWindowTapped", id)
        }

        fun onBackButtonTapped() {
            this.channel.invokeMethod("backButtonTapped", null)
        }

        fun getAssetFileDecriptor(asset: String): AssetFileDescriptor {
            val assetManager = registrar.context().getAssets()
            val key = registrar.lookupKeyForAsset(asset)
            return assetManager.openFd(key)
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result): Unit {
        when {
            call.method == "setApiKey" -> result.success(false)
            call.method == "show" -> {
                val code = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(activity)
                if (GoogleApiAvailability.getInstance().showErrorDialogFragment(activity, code, REQUEST_GOOGLE_PLAY_SERVICES)) {
                    return
                }
                val mapOptions = call.argument<Map<String, Any>>("mapOptions")
                val cameraDict = mapOptions["cameraPosition"] as Map<String, Any>
                initialCameraPosition = getCameraPosition(cameraDict)
                toolbarActions = getToolbarActions(call.argument<List<Map<String, Any>>>("actions"))
                showUserLocation = mapOptions["showUserLocation"] as Boolean
                showMyLocationButton = mapOptions["showMyLocationButton"] as Boolean
                showCompassButton = mapOptions["showCompassButton"] as Boolean
                hideToolbar = mapOptions["hideToolbar"] as Boolean
                mapTitle = mapOptions["title"] as String

                if (mapOptions["mapViewType"] != null) {
                    val mappedMapType: Int? = mapTypeMapping.get(mapOptions["mapViewType"]);
                    if (mappedMapType != null) mapViewType = mappedMapType;
                }

                val intent = Intent(activity, MapActivity::class.java)
                activity.startActivity(intent)
                result.success(true)
                return
            }
            call.method == "dismiss" -> {
                mapActivity?.finish()
                result.success(true)
                return
            }
            call.method == "getZoomLevel" -> {
                val zoom = mapActivity?.zoomLevel ?: 0.0.toFloat()
                result.success(zoom)
            }
            call.method == "getCenter" -> {
                val center = mapActivity?.target ?: LatLng(0.0, 0.0)
                result.success(mapOf("latitude" to center.latitude,
                        "longitude" to center.longitude))
            }
            call.method == "setCamera" -> {
                handleSetCamera(call.arguments as Map<String, Any>)
                result.success(true)
            }
            call.method == "zoomToAnnotations" -> {
                handleZoomToAnnotations(call.arguments as Map<String, Any>)
                result.success(true)
            }
            call.method == "zoomToPolylines" -> {
                handleZoomToPolylines(call.arguments as Map<String, Any>)
                result.success(true)
            }
            call.method == "zoomToPolygons" -> {
                handleZoomToPolygons(call.arguments as Map<String, Any>)
                result.success(true)
            }
            call.method == "zoomToFit" -> {
                mapActivity?.zoomToFit(call.arguments as Int)
                result.success(true)
            }
            call.method == "getVisibleMarkers" -> {
                val visibleMarkerIds = mapActivity?.visibleMarkers ?: emptyList()
                result.success(visibleMarkerIds)
            }
            call.method == "clearAnnotations" -> {
                mapActivity?.clearMarkers()
                result.success(true)
            }
            call.method == "setAnnotations" -> {
                handleSetAnnotations(call.arguments as List<Map<String, Any>>)
                result.success(true)
            }
            call.method == "addAnnotation" -> {
                handleAddAnnotation(call.arguments as Map<String, Any>)
            }
            call.method == "removeAnnotation" -> {
                handleRemoveAnnotation(call.arguments as Map<String, Any>)
            }
            call.method == "getVisiblePolylines" -> {
                val visiblePolylineIds = mapActivity?.visiblePolyline ?: emptyList()
                result.success(visiblePolylineIds)
            }
            call.method == "clearPolylines" -> {
                mapActivity?.clearPolylines()
                result.success(true)
            }
            call.method == "setPolylines" -> {
                handleSetPolylines(call.arguments as List<Map<String, Any>>)
                result.success(true)
            }
            call.method == "addPolyline" -> {
                handleAddPolyline(call.arguments as Map<String, Any>)
            }
            call.method == "removePolyline" -> {
                handleRemovePolyline(call.arguments as Map<String, Any>)
            }
            call.method == "getVisiblePolygons" -> {
                val visiblePolygonIds = mapActivity?.visiblePolygon ?: emptyList()
                result.success(visiblePolygonIds)
            }
            call.method == "clearPolygons" -> {
                mapActivity?.clearPolygons()
                result.success(true)
            }
            call.method == "setPolygons" -> {
                handleSetPolygons(call.arguments as List<Map<String, Any>>)
                result.success(true)
            }
            call.method == "addPolygon" -> {
                handleAddPolygon(call.arguments as Map<String, Any>)
            }
            call.method == "removePolygon" -> {
                handleRemovePolygon(call.arguments as Map<String, Any>)
            }
            else -> result.notImplemented()
        }
    }

    fun handleSetCamera(map: Map<String, Any>) {
        val lat = map["latitude"] as Double
        val lng = map["longitude"] as Double
        val zoom = map["zoom"] as Double
        val bearing = map["bearing"] as Double
        val tilt = map["tilt"] as Double
        mapActivity?.setCamera(LatLng(lat, lng), zoom.toFloat(), bearing.toFloat(), tilt.toFloat())
    }

    fun handleZoomToAnnotations(map: Map<String, Any>) {
        val ids = map["annotations"] as List<String>
        val padding = map["padding"] as Double
        mapActivity?.zoomToAnnotations(ids, padding.toFloat())
    }

    fun handleZoomToPolylines(map: Map<String, Any>) {
        val ids = map["polylines"] as List<String>
        val padding = map["padding"] as Double
        mapActivity?.zoomToPolylines(ids, padding.toFloat())
    }

    fun handleZoomToPolygons(map: Map<String, Any>) {
        val ids = map["polygons"] as List<String>
        val padding = map["padding"] as Double
        mapActivity?.zoomToPolygons(ids, padding.toFloat())
    }

    fun handleSetAnnotations(annotations: List<Map<String, Any>>) {
        val mapAnnoations = ArrayList<MapAnnotation>()
        for (a in annotations) {
            val mapAnnotation = MapAnnotation.fromMap(a)
            if (mapAnnotation != null) {
                mapAnnoations.add(mapAnnotation)
            }
        }
        mapActivity?.setAnnotations(mapAnnoations)
    }

    fun handleAddAnnotation(map: Map<String, Any>) {
        MapAnnotation.fromMap(map)?.let {
            mapActivity?.addMarker(it)
        }
    }

    fun handleRemoveAnnotation(map: Map<String, Any>) {
        MapAnnotation.fromMap(map)?.let {
            mapActivity?.removeMarker(it)
        }
    }

    fun handleSetPolylines(polylines: List<Map<String, Any>>) {
        val mapPolylines = ArrayList<MapPolyline>()
        for (a in polylines) {
            val mapPolyline = MapPolyline.fromMap(a)
            if (mapPolyline != null) {
                mapPolylines.add(mapPolyline)
            }
        }
        mapActivity?.setPolylines(mapPolylines)
    }

    fun handleAddPolyline(map: Map<String, Any>) {
        MapPolyline.fromMap(map)?.let {
            mapActivity?.addPolyline(it)
        }
    }

    fun handleRemovePolyline(map: Map<String, Any>) {
        MapPolyline.fromMap(map)?.let {
            mapActivity?.removePolyline(it)
        }
    }

    fun handleSetPolygons(polygons: List<Map<String, Any>>) {
        val mapPolygons = ArrayList<MapPolygon>()
        for (a in polygons) {
            val mapPolygon = MapPolygon.fromMap(a)
            if (mapPolygon != null) {
                mapPolygons.add(mapPolygon)
            }
        }
        mapActivity?.setPolygons(mapPolygons)
    }

    fun handleAddPolygon(map: Map<String, Any>) {
        MapPolygon.fromMap(map)?.let {
            mapActivity?.addPolygon(it)
        }
    }

    fun handleRemovePolygon(map: Map<String, Any>) {
        MapPolygon.fromMap(map)?.let {
            mapActivity?.removePolygon(it)
        }
    }
}

